Animatediff Proposal

DN6 opened the pull request:
# What does this PR do?
Adds the [AnimateDiff](https://animatediff.github.io/) model to `diffusers`. 

This PR:

1. Introduces a dedicated `UNetMotionModel` and motion blocks that allow using existing 2D UNet models with the Motion Modules introduced in AnimateDiff. 

The approach taken by AnimateDiff is slightly different to a 3D UNet.  No temporal convs are applied to the inputs and the motion module weights are saved separately from the 2D UNet weights. This is slightly similar to a ControlNet/Adapter approach so IMO it would be better to create a dedicated model and blocks to support this type of functionality, rather than adapt the existing 3D UNets. 

2. Adds a `MotionAdapter` module that acts as a dedicated container for saving and loading motion module weights from the hub. 

Proposed API

```python
# Loading motion module into Pipeline
from diffusers import MotionAdapter, AnimateDiffPipeline

motion_adapter = MotionAdapter.from_pretrained("<path to saved motion modules>")
pipe = AnimateDiffPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", motion_adapter=motion_adapter)

# Calling pipe.unet should return an UNetMotionModel object

# Create brand new UNetMotionModel with all random weights
unet = UNetMotionModel()

# Load from an existing 2D UNet and MotionAdapter
unet2D = UNet2DConditionModel.from_pretrained("...")
motion_adapter = MotionAdapter.from_pretrained("...")

unet_motion = UNetMotionModel.from_unet2d(unet2D, motion_adapter: Optional = None)

# Or load motion module after init
unet_motion.load_motion_modules(motion_adapter)

# Save only motion modules
unet_motion.save_motion_module(<path to save model>, push_to_hub=True)

# Save all weights to a single model repo (Including UNet weights) 
unet_motion.save_pretrained()

# Load fused models (Where the motion weights are saved along with the UNet weights in a single repo)
unet_motion = UNetMotionModel.from_pretrained("<path to model>") 
``` 

TODO: 

1. Test forward pass of UNetMotionModel to verify outputs match the original model
2. Test pipeline outputs for the same prompts defined in the original model repo
3. Support loading Motion LoRAs  
4. Add documentation
5. Add tests
6. Add support for training/fine tuning AnimateDiff models. 

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Core library:

- Schedulers: @williamberman and @patrickvonplaten
- Pipelines:  @patrickvonplaten and @sayakpaul
- Training examples: @sayakpaul and @patrickvonplaten
- Docs: @stevhliu and @yiyixuxu
- JAX and MPS: @pcuenca
- Audio: @sanchit-gandhi
- General functionalities: @patrickvonplaten and @sayakpaul

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- transformers: [different repo](https://github.com/huggingface/transformers)
- safetensors: [different repo](https://github.com/huggingface/safetensors)

-->

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_5413). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten commented on src/diffusers/models/embeddings.py:
```diff
@@ -249,6 +249,34 @@ def forward(self, x):
         return out
 
 
+class PositionalEmbedding(nn.Module):
```
Why can't we leverage: https://github.com/huggingface/diffusers/blob/33293ed50470f681ad8c704e325b2960dc66cfce/src/diffusers/models/embeddings.py#L26 here?

patrickvonplaten replied:
These are sinoidal position embeddings, let's maybe name them differently

patrickvonplaten commented on src/diffusers/models/unet_motion_blocks.py:
```diff
@@ -0,0 +1,1234 @@
+from typing import Any, Dict, Optional, Tuple
```
Not too happy about creating a new file here. Most of the code should go either into the unet_3d_blocks or 2d_blocks code

DN6 replied:
I think I can move it into the 2D blocks, since all the motion blocks are mods of the 2D ones.

patrickvonplaten commented on src/diffusers/models/unet_motion_blocks.py:
```diff
@@ -0,0 +1,1234 @@
+    raise ValueError(f"{up_block_type} does not exist.")
+
+
+class MotionAttnProcessor(nn.Module):
```
These attn processors are very specific to motion so let's add them to `unet_modition_model.py`

DN6 replied:
Hmm so there might be a circular dependency issue here. 

The UNet Motion blocks have Motion Modules that use these attention processors. So you would end up in a situation where importing the blocks into `unet_motion_model.py` would require importing the model into the blocks. 

Could we create a separate `motion.py` file for the MotionModules, MotionAdapter, Motion Attn Processors and move the UNet Motion blocks into the 2D Unet blocks file?

patrickvonplaten replied:
Ah I see, I'd then maybe move the attn processors to `attention.py` and the blocks into unet_3d_blocks

patrickvonplaten commented on src/diffusers/models/unet_motion_blocks.py:
```diff
@@ -0,0 +1,1234 @@
+            )
+
+
+class MotionAdapter(ModelMixin, ConfigMixin):
```
This should be in its own file. Because it's very special I'd probably just add it to the `animatediff` folder

patrickvonplaten commented on src/diffusers/models/unet_motion_model.py:
```diff
@@ -0,0 +1,813 @@
+
+        self.mid_block.motion_modules.load_state_dict(motion_adapter.mid_block.motion_modules.state_dict())
+
+    def save_motion_modules(
```
Nice!

patrickvonplaten commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -0,0 +1,665 @@
+        ],
+    ):
+        super().__init__()
+        unet = UNetMotionModel.from_unet2d(unet, motion_adapter)
```
nice that's clean!

patrickvonplaten commented on src/diffusers/models/unet_motion_model.py:
```diff
@@ -0,0 +1,813 @@
+
+        model = cls.from_config(config)
+
+        if load_weights:
```
I think this should always be True and not exist as a function argument. Why would one **not** load the unet parameters?

DN6 replied:
Hmm it is unlikely. I set it to be an argument in order to enable training a full model (UNet + Motion Modules). It follows the ControlNet design
https://github.com/huggingface/diffusers/blob/c7617e482a522173ea6f922223aa010058552af8/src/diffusers/models/controlnet.py#L425

a-r-r-o-w commented on src/diffusers/models/attention.py:
```diff
@@ -275,6 +275,215 @@ def forward(
+        dim: int,
+        num_attention_heads: int,
+        attention_head_dim: int,
+        dropout=0.0,
```
```suggestion
        dropout: float = 0.0,
```

a-r-r-o-w commented on src/diffusers/models/attention.py:
```diff
@@ -275,6 +275,215 @@ def forward(
+        norm_type: str = "layer_norm",
+        final_dropout: bool = False,
+        attention_type: str = "default",
+        max_seq_length=32,
```
```suggestion
        max_seq_length: int = 32,
```

a-r-r-o-w commented on src/diffusers/models/attention.py:
```diff
@@ -275,6 +275,215 @@ def forward(
+            Whether to apply a final dropout after the last feed-forward layer.
+        attention_type (`str`, *optional*, defaults to `"default"`):
+            The type of attention to use. Can be `"default"` or `"gated"` or `"gated-text-image"`.
+    """
```
```suggestion
        max_seq_length (`int`, *optional*, defaults to 32):
            Maximum sequence length to apply positional embeddings.
    """
```

a-r-r-o-w commented on src/diffusers/models/transformer_temporal.py:
```diff
@@ -183,3 +183,151 @@ def forward(
+            Configure if the `TransformerBlock` attention should contain a bias parameter.
+        double_self_attention (`bool`, *optional*):
+            Configure if each `TransformerBlock` should contain two self-attention layers.
+    """
```
```suggestion
        max_seq_length (`int`, *optional*, defaults to 32):
            Maximum sequence length to apply positional embeddings.
    """
```

a-r-r-o-w reviewed (approved):
Great to see Animatediff being added. Awesome work, thanks! I realized there are many typing/doc improvements that can be made here so will do those in a seperate PR post merge in unet2d/3d.

DN6 commented:
@patrickvonplaten based on the initial feedback I made the following changes

1. Moved all UNet motion blocks into `unet3d_blocks.py` 
2. Added a `TransfomerTemporalPosEmbedBlock` that handles adding the position embeddings before each attention layer and remove the custom attn processors for motion
3. Added a `TransformerTemporalMotionModel` to `transformer_temporal` that uses this new block and follows the reshaping/norm steps of the original implementation exactly. 
4. Arguments that are specific to the `TransformerTemporalMotionModel` in the `get_up_block` and `get_down_block` functions  the more generic prefix of `temporal` rather than `motion`
5. Move `MotionAdapter` and other motion modules into `unet_motion_model.py`

patrickvonplaten commented on src/diffusers/models/transformer_temporal.py:
```diff
@@ -183,3 +183,153 @@ def forward(
         return TransformerTemporalModelOutput(sample=output)
+
+
+class TransformerTemporalMotionModel(ModelMixin, ConfigMixin):
```
Let's try to merge into Transformer2DModel instead

DN6 replied:
Hmm I think it can be merged into the `TransformerTemporalModel`.

DN6 commented:
@patrickvonplaten This is ready for another review. 

TODO:
1. Finish pipeline test
2. Add docs page.

patrickvonplaten commented on src/diffusers/models/attention.py:
```diff
@@ -115,6 +115,8 @@ def __init__(
         norm_type: str = "layer_norm",
         final_dropout: bool = False,
         attention_type: str = "default",
+        use_positional_embedding: bool = False,
```
```suggestion
        position_embeddings: Optional[str] = None,  # 'sinusoidal'
```

patrickvonplaten replied:
String attributes are easier to expand in the future

patrickvonplaten commented on src/diffusers/models/attention.py:
```diff
@@ -115,6 +115,8 @@ def __init__(
         final_dropout: bool = False,
         attention_type: str = "default",
+        use_positional_embedding: bool = False,
+        max_seq_length: Optional[int] = None,
```
```suggestion
        num_positition_embeddings: Optional[int] = None,
```

patrickvonplaten commented on src/diffusers/models/attention.py:
```diff
@@ -128,6 +130,14 @@ def __init__(
                 f" define `num_embeds_ada_norm` if setting `norm_type` to {norm_type}."
             )
 
+        if use_positional_embedding and (max_seq_length is None):
```
```suggestion
        if position_embeddings == "sinusoidal" and (num_positition_embeddings is None):
```

patrickvonplaten commented on src/diffusers/models/attention.py:
```diff
@@ -115,6 +115,8 @@ def __init__(
         norm_type: str = "layer_norm",
         final_dropout: bool = False,
         attention_type: str = "default",
+        use_positional_embedding: bool = False,
```
Let's not forget doc strings here

patrickvonplaten commented on src/diffusers/models/embeddings.py:
```diff
@@ -249,6 +249,30 @@ def forward(self, x):
         return out
 
 
+class PositionalEmbedding(nn.Module):
```
```suggestion
class SinusoidalEmbeddings:
```

patrickvonplaten commented on src/diffusers/models/embeddings.py:
```diff
@@ -249,6 +249,30 @@ def forward(self, x):
+        super().__init__()
+        pe = get_1d_sincos_pos_embed_from_grid(embed_dim, torch.arange(32))
+        pe = torch.tensor(pe, dtype=torch.float32).unsqueeze(0)
+        self.register_buffer("pe", pe)
```
```suggestion
```

DN6 replied:
@patrickvonplaten is there a specific reason to avoid using `self.register_buffer`? It makes changing dtypes and devices much cleaner IMO.

patrickvonplaten replied:
Yeah I'd usually just do this in the forward method, but ok to leave it. The reason I don't like to register as a buffer is because it'll save the sinus position embeddings as part of the weights, making them unnecessarily heavy. But ok to leave!

patrickvonplaten commented on src/diffusers/models/embeddings.py:
```diff
@@ -249,6 +249,30 @@ def forward(self, x):
+    def __init__(self, embed_dim: int, max_seq_length: int = 32):
+        super().__init__()
+        pe = get_1d_sincos_pos_embed_from_grid(embed_dim, torch.arange(32))
+        pe = torch.tensor(pe, dtype=torch.float32).unsqueeze(0)
```
```suggestion
        self.pe = torch.tensor(pe, dtype=torch.float32).unsqueeze(0)
```

patrickvonplaten commented on src/diffusers/models/embeddings.py:
```diff
@@ -249,6 +249,30 @@ def forward(self, x):
+    """
+
+    def __init__(self, embed_dim: int, max_seq_length: int = 32):
+        super().__init__()
```
```suggestion
```

patrickvonplaten commented on src/diffusers/models/attention.py:
```diff
@@ -128,6 +130,14 @@ def __init__(
+        if use_positional_embedding and (max_seq_length is None):
+            raise ValueError("If `use_positional_embedding` is set to `True`, `max_sequence_length` must be defined.")
+
+        if use_positional_embedding:
```
```suggestion
        if position_embeddings == "sinusoidal":
```

patrickvonplaten commented on src/diffusers/models/attention.py:
```diff
@@ -128,6 +130,14 @@ def __init__(
+            raise ValueError("If `use_positional_embedding` is set to `True`, `max_sequence_length` must be defined.")
+
+        if use_positional_embedding:
+            self.pos_embed = PositionalEmbedding(dim, max_seq_length=max_seq_length)
```
```suggestion
            self.pos_embed = SinusoidalEmbedding(dim, max_seq_length=max_seq_length)
```

patrickvonplaten commented on src/diffusers/models/transformer_temporal.py:
```diff
@@ -149,13 +161,28 @@ def forward(
 
-        hidden_states = self.norm(hidden_states)
-        hidden_states = hidden_states.permute(0, 3, 4, 2, 1).reshape(batch_size * height * width, num_frames, channel)
+        torch.save(hidden_states, "hs-notframewise.pt")
```
```suggestion
```

patrickvonplaten commented on src/diffusers/models/transformer_temporal.py:
```diff
@@ -149,13 +161,28 @@ def forward(
+        torch.save(hidden_states, "hs-notframewise.pt")
 
         hidden_states = self.proj_in(hidden_states)
+        encoder_hidden_states = encoder_hidden_states if self.use_cross_attention else None
```
```suggestion
        encoder_hidden_states = encoder_hidden_states if self.use_cross_attention else None
```
do we need this?

patrickvonplaten replied:
I still don't understand why we need this. If `cross_attention_dim` is None then why do we have to manually set encodre_hidden_states to `None`. This looks more like a hacky bug correction. Why do we pass `encoder_hidden_states` in the first place if we don't have cross attention?

DN6 replied:
@patrickvonplaten 
We can omit sending the encoder hidden state to this block from the higher level blocks.  It just means the MotionModules in the UNetMotionModel cannot support cross attention at all.

patrickvonplaten replied:
We should only add options that are needed to run the official animate diff checkpoints, all possible customizations that the user could try out should not be added (only if it becomes necessary)

patrickvonplaten replied:
Here if the motion modules all have `temporal_cross_attention_dim` always set to None, the let's **not** give the possibility to customize it as this unnecessarily bloats the code

patrickvonplaten commented on src/diffusers/models/transformer_temporal.py:
```diff
@@ -149,13 +161,28 @@ def forward(
+        if not self.apply_framewise_group_norm:
+            hidden_states = self.norm(hidden_states)
+            hidden_states = hidden_states[None, :].reshape(batch_size, num_frames, channel, height, width)
+            hidden_states = hidden_states.permute(0, 3, 4, 1, 2).reshape(
+                batch_size * height * width, num_frames, channel
+            )
```
```suggestion
```

patrickvonplaten commented on src/diffusers/models/transformer_temporal.py:
```diff
@@ -149,13 +161,28 @@ def forward(
+            hidden_states = hidden_states.permute(0, 2, 1, 3, 4)
+
+            hidden_states = self.norm(hidden_states)
+            hidden_states = hidden_states.permute(0, 3, 4, 2, 1).reshape(
+                batch_size * height * width, num_frames, channel
+            )
```
```suggestion
        hidden_states = hidden_states.permute(0, 3, 4, 2, 1).reshape(
            batch_size * height * width, num_frames, channel
        )
```

patrickvonplaten commented on src/diffusers/models/transformer_temporal.py:
```diff
@@ -55,6 +55,12 @@ class TransformerTemporalModel(ModelMixin, ConfigMixin):
             Configure if the `TransformerBlock` attention should contain a bias parameter.
```
Nice that works!

patrickvonplaten commented on src/diffusers/models/unet_3d_blocks.py:
```diff
@@ -74,6 +84,53 @@ def get_down_block(
         )
+    if down_block_type == "DownBlockMotion":
+        return DownBlockMotion(
+            num_layers=num_layers,
+            in_channels=in_channels,
+            out_channels=out_channels,
+            temb_channels=temb_channels,
+            add_downsample=add_downsample,
+            resnet_eps=resnet_eps,
+            resnet_act_fn=resnet_act_fn,
+            resnet_groups=resnet_groups,
+            downsample_padding=downsample_padding,
+            resnet_time_scale_shift=resnet_time_scale_shift,
+            temporal_norm_num_groups=temporal_norm_num_groups,
+            temporal_cross_attention_dim=temporal_cross_attention_dim,
+            temporal_num_attention_heads=temporal_num_attention_heads,
+            temporal_attention_bias=temporal_attention_bias,
+            temporal_activation_fn=temporal_activation_fn,
+            temporal_max_seq_length=temporal_max_seq_length,
```
```suggestion
            num_layers=num_layers,
            in_channels=in_channels,
            out_channels=out_channels,
            temb_channels=temb_channels,
            add_downsample=add_downsample,
            resnet_eps=resnet_eps,
            resnet_act_fn=resnet_act_fn,
            resnet_groups=resnet_groups,
            downsample_padding=downsample_padding,
            resnet_time_scale_shift=resnet_time_scale_shift,
            temporal_norm_num_groups=temporal_norm_num_groups,
            temporal_cross_attention_dim=temporal_cross_attention_dim,
            temporal_num_attention_heads=temporal_num_attention_heads,
            temporal_attention_bias=temporal_attention_bias,
            temporal_activation_fn=temporal_activation_fn,
            temporal_max_seq_length=temporal_max_seq_length,
```
Do we really need to pass all these arguments? Let's maybe try to set more defaults that hold true for all motion modules. I'm a bit surprised that we need to allow for that much customization right at the beginning.

patrickvonplaten commented on src/diffusers/models/unet_3d_blocks.py:
```diff
@@ -724,3 +835,825 @@ def forward(self, hidden_states, res_hidden_states_tuple, temb=None, upsample_si
+        temporal_num_attention_heads=8,
+        temporal_attention_bias=False,
+        temporal_activation_fn="geglu",
+        temporal_max_seq_length=32,
```
Do we need all these arguments? Can you set more arguments to sensible values that we don't allow the user to change?

patrickvonplaten commented on src/diffusers/models/unet_motion_model.py:
```diff
@@ -0,0 +1,939 @@
+
+        def fn_recursive_add_processors(name: str, module: torch.nn.Module, processors: Dict[str, AttentionProcessor]):
+            if hasattr(module, "get_processor"):
+                processors[f"{name}.processor"] = module.get_processor(return_deprecated_lora=True)
```
```suggestion
                processors[f"{name}.processor"] = module.get_processor()
```
Let's make sure to not copy deprecated functionality

patrickvonplaten commented on src/diffusers/models/unet_motion_model.py:
```diff
@@ -0,0 +1,939 @@
+
+        return processors
+
+    # Copied from diffusers.models.unet_2d_condition.UNet2DConditionModel.set_attention_slice
+    def set_attention_slice(self, slice_size):
+        r"""
+        Enable sliced attention computation.
+
+        When this option is enabled, the attention module splits the input tensor in slices to compute attention in
+        several steps. This is useful for saving some memory in exchange for a small decrease in speed.
+
+        Args:
+            slice_size (`str` or `int` or `list(int)`, *optional*, defaults to `"auto"`):
+                When `"auto"`, input to the attention heads is halved, so attention is computed in two steps. If
+                `"max"`, maximum amount of memory is saved by running only one slice at a time. If a number is
+                provided, uses as many slices as `attention_head_dim // slice_size`. In this case, `attention_head_dim`
+                must be a multiple of `slice_size`.
+        """
+        sliceable_head_dims = []
+
```
```suggestion
```

Let's not add this anymore. I'm sure that 99% of users by now use flash attention and when doing so attention slicing gives to speed ups

patrickvonplaten commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -0,0 +1,653 @@
+        callback_steps: Optional[int] = 1,
+        cross_attention_kwargs: Optional[Dict[str, Any]] = None,
+        clip_skip: Optional[int] = None,
+        **kwargs,
```
```suggestion
```
Let's be careful with adding kwargs and try to avoid it

patrickvonplaten reviewed (commented):
General structure looks very nice now! Left some final things that could be improved.

1. Let's make sure to never copy-paste deprecated functionality
2. When adding new blocks, classes that had lots of customization in the original codebase, we should not copy all of these customization abilities. Instead we should only allow parameter to be customizable that differ between publicized checkpoints. Especially the motion blocks look a bit too scary here. Can we try to reduce the number of init parameters to less than half of what is there now? 
3. When adding new config parameters (such as `use_position_embeddings`) it's always better to add string configs instead of booleans so that one can extend those in the future.

I think the things above are quite easy to change and then we should be good for merge

DN6 commented:
@patrickvonplaten I think this PR is ready for a final review. Just a summary of the changes 

1. Switched back to the original PostionalEmbeddings implementation from the authoer. Checkpoint that use our implementation of the Positional Embedding do not produce coherent results. I think it's because the way that sin and cos values in the sequence are interleaved in the original vs concatenated in ours. 

2. Only left the following temporal/motion arguments in `get_down_block` and `get_up_block`

```
temporal_num_attention_heads=8,
temporal_cross_attention_dim=None,
temporal_max_seq_length=32,
```

`temporal_attention_heads` needs to be passed in because the `DownBlockMotion`/`UpBlockMotion` blocks don't have a `num_attention_heads` argument so there's no other way to configure the motion modules in these blocks. 

`temporal_cross_attention_dim` and `temporal_max_seq_length` allow the user to train a motion modules with cross attention, and adjust the maximum video length the model can train on. 
 
 3. I remove the different group norm options in transformer temporal. We can just use the approach that's already exists. It doesn't make any difference to the output.

patrickvonplaten commented on src/diffusers/models/attention.py:
```diff
@@ -128,6 +135,16 @@ def __init__(
                 f" define `num_embeds_ada_norm` if setting `norm_type` to {norm_type}."
             )
 
+        if positional_embeddings and (num_positional_embeddings is None):
```
```suggestion
        if positional_embeddings is not None and num_positional_embeddings is None:
```

patrickvonplaten commented on src/diffusers/models/transformer_temporal.py:
```diff
@@ -67,22 +69,23 @@ def __init__(
         norm_num_groups: int = 32,
         cross_attention_dim: Optional[int] = None,
         attention_bias: bool = False,
-        sample_size: Optional[int] = None,
```
We should leave those for backward compatibility

patrickvonplaten replied:
Someone might rely on model.config.sample_size

patrickvonplaten commented on src/diffusers/models/unet_3d_condition.py:
```diff
@@ -366,6 +341,31 @@ def fn_recursive_set_attention_slice(module: torch.nn.Module, slice_size: List[i
         for module in self.children():
             fn_recursive_set_attention_slice(module, reversed_slice_size)
 
+    @property
```
Why move it down?

DN6 replied:
I think this was a mistake. Will fix.

patrickvonplaten commented on src/diffusers/models/unet_3d_condition.py:
```diff
@@ -502,7 +502,6 @@ def forward(
         sample: torch.FloatTensor,
         timestep: Union[torch.Tensor, float, int],
         encoder_hidden_states: torch.Tensor,
-        class_labels: Optional[torch.Tensor] = None,
```
Why do we delete this? This is backward breaking.

If `class_labels` is not needed, let's add a `kwarg`  that gracefully deprecates `class_labels`

patrickvonplaten commented on src/diffusers/models/unet_motion_model.py:
```diff
@@ -0,0 +1,882 @@
+        norm_num_groups: Optional[int] = 32,
+        norm_eps: float = 1e-5,
+        cross_attention_dim: int = 1280,
+        attention_head_dim: Union[int, Tuple[int]] = 8,
```
```suggestion
```

patrickvonplaten replied:
Let's correct what we incorrectly did before

patrickvonplaten commented on src/diffusers/models/unet_motion_model.py:
```diff
@@ -0,0 +1,882 @@
+                "At the moment it is not possible to define the number of attention heads via `num_attention_heads` because of a naming issue as described in https://github.com/huggingface/diffusers/issues/2011#issuecomment-1547958131. Passing `num_attention_heads` will only be supported in diffusers v0.19."
+            )
+
+        # If `num_attention_heads` is not defined (which is the case for most models)
+        # it will default to `attention_head_dim`. This looks weird upon first reading it and it is.
+        # The reason for this behavior is to correct for incorrectly named variables that were introduced
+        # when this library was created. The incorrect naming was only discovered much later in https://github.com/huggingface/diffusers/issues/2011#issuecomment-1547958131
+        # Changing `attention_head_dim` to `num_attention_heads` for 40,000+ configurations is too backwards breaking
+        # which is why we correct for the naming here.
+        num_attention_heads = num_attention_heads or attention_head_dim
+
```
```suggestion
```

patrickvonplaten commented on src/diffusers/models/unet_motion_model.py:
```diff
@@ -0,0 +1,882 @@
+        norm_eps: float = 1e-5,
+        cross_attention_dim: int = 1280,
+        attention_head_dim: Union[int, Tuple[int]] = 8,
+        use_linear_projection: bool = False,
```
Do we need this? Does the model use a linear projection?

DN6 replied:
It's passed in to create the UNet2D blocks.

patrickvonplaten commented on src/diffusers/models/unet_motion_model.py:
```diff
@@ -0,0 +1,882 @@
+        for module in self.children():
+            fn_recursive_feed_forward(module, None, 0)
+
+    def set_default_attn_processor(self):
```
add copied from

patrickvonplaten commented on src/diffusers/models/unet_motion_model.py:
```diff
@@ -0,0 +1,882 @@
+        for name, module in self.named_children():
+            fn_recursive_attn_processor(name, module, processor)
+
+    def enable_forward_chunking(self, chunk_size=None, dim=0):
```
add copied from

patrickvonplaten commented on src/diffusers/models/unet_motion_model.py:
```diff
@@ -0,0 +1,882 @@
+        for module in self.children():
+            fn_recursive_feed_forward(module, chunk_size, dim)
+
+    def disable_forward_chunking(self):
```
add copied from

patrickvonplaten commented on src/diffusers/models/unet_motion_model.py:
```diff
@@ -0,0 +1,882 @@
+
+        self.set_attn_processor(processor, _remove_lora=True)
+
+    def _set_gradient_checkpointing(self, module, value=False):
```
add copied from

DN6 replied:
Here the checkpointing is happening on the motion blocks. I don't think copied from applies?

patrickvonplaten commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -0,0 +1,651 @@
+        tokenizer (`CLIPTokenizer`):
+            A [`~transformers.CLIPTokenizer`] to tokenize text.
+        unet ([`UNet3DConditionModel`]):
+            A [`UNet3DConditionModel`] to denoise the encoded video latents.
```
motion adapter doc string is missing

patrickvonplaten commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -0,0 +1,651 @@
+        self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1)
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor)
+
+    def encode_prompt(
```
can you add a copied from here or not?

DN6 replied:
Hmm I could change the argument `num_videos_per_prompt

patrickvonplaten commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -0,0 +1,651 @@
+        """Disables the FreeU mechanism if enabled."""
+        self.unet.disable_freeu()
+
+    def prepare_extra_step_kwargs(self, generator, eta):
```
add copied from

patrickvonplaten commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -0,0 +1,651 @@
+                    f" {negative_prompt_embeds.shape}."
+                )
+
+    def prepare_latents(
```
can't you add a copied from here from https://github.com/huggingface/diffusers/blob/d1eb14bc357a179638be003cd61795f5dc2045f8/src/diffusers/pipelines/text_to_video_synthesis/pipeline_text_to_video_synth.py#L457

patrickvonplaten reviewed (commented):
I've added some final comments. It would be great if you could go over them and then I think we can merge tomorrow!

patrickvonplaten commented on src/diffusers/models/transformer_temporal.py:
```diff
@@ -67,7 +71,6 @@ def __init__(
         num_attention_heads: int = 16,
         attention_head_dim: int = 88,
         in_channels: Optional[int] = None,
-        out_channels: Optional[int] = None,
```
why do we delete this here?

patrickvonplaten replied:
Let's keep it for backwards compatibility reasons

patrickvonplaten commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -0,0 +1,688 @@
+        self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1)
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor)
+
+    def encode_prompt(
```
```suggestion
    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.encode_prompt with num_images_per_prompt -> num_videos_per_prompt
    def encode_prompt(
```

patrickvonplaten commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -0,0 +1,688 @@
+
+        output_frames = []
+
+        # decode frame by frame to avoid OOM
```
The user can decide to do this by enabling vae slicing we shouldn't do it ourselves. Can you just copy from this function: https://github.com/huggingface/diffusers/blob/49255f2d94177966f4ba9623d6eddfaebd359e9f/src/diffusers/pipelines/text_to_video_synthesis/pipeline_text_to_video_synth.py#L358

patrickvonplaten commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -0,0 +1,688 @@
+
+        return prompt_embeds, negative_prompt_embeds
+
+    def decode_latents(self, latents):
```
```suggestion
     # Copied from diffusers.pipelines.text_to_video_synthesis/pipeline_text_to_video_synth.TextToVideoSDPipeline.decode_latents
    def decode_latents(self, latents):
```

DN6 replied:
I think this is different from the TextToVideoSDPipeline decode_latents method. It decodes each latent frame one by one to avoid OOM.

patrickvonplaten commented on docs/source/en/api/pipelines/animatediff.md:
```diff
@@ -0,0 +1,90 @@
+adapter = MotionAdapter.from_pretrained("guoyww/animatediff-motion-adapter-v1-5-2")
+# load SD 1.5 based finetuned model
+pipe = AnimateDiffPipeline.from_pretrained("frankjoshua/toonyou_beta6", motion_adapter=adapter)
+pipe.scheduler = DDIMScheduler(
```
Let's load the DDIM scheduler from somewhere instead of forcing the user to do all this

patrickvonplaten commented on docs/source/en/api/pipelines/animatediff.md:
```diff
@@ -0,0 +1,90 @@
+# Load the motion adapter
+adapter = MotionAdapter.from_pretrained("guoyww/animatediff-motion-adapter-v1-5-2")
+# load SD 1.5 based finetuned model
+pipe = AnimateDiffPipeline.from_pretrained("frankjoshua/toonyou_beta6", motion_adapter=adapter)
```
this model has no model card, can we maybe use another model as the main example?

patrickvonplaten commented on docs/source/en/api/pipelines/animatediff.md:
```diff
@@ -0,0 +1,90 @@
+    timestep_spacing="linspace",
+)
+# enable memory savings
+pipe.enable_model_cpu_offload()
```
```suggestion
pipe.enable_model_cpu_offload()
pipe.enable_vae_slicing()
```

patrickvonplaten commented on docs/source/en/api/pipelines/animatediff.md:
````diff
@@ -0,0 +1,90 @@
+)
+frames = output.frames[0]
+export_to_gif(frames, "animation.gif")
+```
````
Let's show a GIF here no?
